server/protocols/jmap/src/main/java/org/apache/james/jmap/event/PropagateLookupRightListener.java

```diff
@@ -0,0 +1,130 @@
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jmap.event;
+
+import java.util.stream.Stream;
+
+import javax.inject.Inject;
+
+import org.apache.james.mailbox.MailboxListener;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.acl.ACLDiff;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.model.MailboxACL;
+import org.apache.james.mailbox.model.MailboxACL.Entry;
+import org.apache.james.mailbox.model.MailboxACL.Right;
+import org.apache.james.mailbox.model.MailboxPath;
+import org.apache.james.mailbox.store.MailboxSessionMapperFactory;
+import org.apache.james.mailbox.store.StoreRightManager;
+import org.apache.james.mailbox.store.event.EventFactory.MailboxRenamedEventImpl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PropagateLookupRightListener implements MailboxListener {
+    private static final Logger LOGGER = LoggerFactory.getLogger(PropagateLookupRightListener.class);
+
+    private final StoreRightManager storeRightManager;
+    private final MailboxSessionMapperFactory mapperFactory;
+
+    @Inject
+    public PropagateLookupRightListener(StoreRightManager storeRightManager, MailboxSessionMapperFactory mapperFactory) {
+        this.storeRightManager = storeRightManager;
+        this.mapperFactory = mapperFactory;
+    }
+
+    @Override
+    public ListenerType getType() {
+        return ListenerType.ONCE;
+    }
+
+    @Override
+    public ExecutionMode getExecutionMode() {
+        return ExecutionMode.ASYNCHRONOUS;
+    }
+
+    @Override
+    public void event(Event event) {
+        MailboxSession mailboxSession = event.getSession();
+
+        if (event instanceof MailboxACLUpdated) {
+            MailboxACLUpdated aclUpdateEvent = (MailboxACLUpdated) event;
+
+            updateLookupRightOnParent(mailboxSession, event.getMailboxPath(), aclUpdateEvent.getAclDiff());
+        } else if (event instanceof MailboxRenamed) {
```

Comment:
renamed involves changing hierarchy ?

Reply:
Yes, 
EG: when we want to rename a mailbox with mailbox path from A.B to C.B --> then we need to update 'l' right on "C" of "B" contain 'l' right. WDYT?

Reply:
It's ok, it's what I was thinking, it's just that it should not be considered as a rename.
We won't change that right now.

Reply:
we decided to not remove it as it does not expose so much, but that's not a perfect solution.
In the future we could "tag" the mailbox as automatically shared and remove it when needed

Reply:
We don't need to remove the 'l' right on old parents. It is in ticket details if I remember well.

Reply:
in the ticket, I can read : 

> IMO we don't need to "remove" lookup right when removing rights on child mailbox

But I have no idea why you think that. Please elaborate.

Reply:
1. The sharing action was performed by the user, I don't see how it hurts to not revert Lookup propagation.
2. The parent mailboxes existence is already  disclosed
3.  Nobody forbids the client to remove that lookup right.
4. That seems like something we can imagine fixing later.
5. I presented it in Sprint planning 33, and it have been around for 2.5 weeks, the evaluated points depends highly on this choice.

Reply:
Ok, great.